dsl/src/lib.rs

```diff
@@ -0,0 +1,175 @@
+}
+
+// within the compiler, these generate tokens for
+// Value
+struct ParsedExpr {
+    t: ExprType,
+    tokens: TokenStream,
+}
+
+fn parse_literal(it: &mut Peekable<IntoIter>) -> ParsedExpr {
+    let token = it.next().expect("failed to parse literal, hit end of string");
+
+    match token {
+        Punct(ref x) => if x.as_char() == '(' {
+            let v = parse_expr(it);
+
+            it.next().expect("expected closing paren, hit end of string");
+            return v
+        },
+
+        // boolean literals
+        Ident(ref x) => match x.to_string().as_str() {
+            "true"  => return ParsedExpr{ t: ExprType::Bool, tokens: quote!{ true }},
+            "false" => return ParsedExpr{ t: ExprType::Bool, tokens: quote!{ false }},
+
+            // unary not
+            "not"   => {
+                let v = parse_literal(it).tokens;
+                return ParsedExpr{ t: ExprType::Bool, tokens: quote!{ Not::new(#v) }}
+            }
+            _ => panic!("error: unknown identifier"),
+        },
+
+        // kinda ugly but we fallthrough basically
+        _ => {},
+    }
+
+    // integer literals
+    if let Ok(Literal::Integer(i)) = Literal::try_from(token) {
+        let v = i.value::<u64>().expect("i don't think this integer fits?");
+        return ParsedExpr{ t: ExprType::Int, tokens: quote!{ #v }}
+    }
+
+    ParsedExpr{ t: ExprType::Nil, tokens: TokenStream::new() }
+}
+
+fn precedence(it: &mut Peekable<IntoIter>) -> Option<(i32, ExprType, &'static str)> {
+    match it.peek() {
+        Some(Punct(x)) => match x.as_char() {
+            // arithmatic
+            '+' => Some((4, ExprType::Int, "Add")),
+            '-' => Some((4, ExprType::Int, "Subtract")),
+            '*' => Some((3, ExprType::Int, "Multiply")),
+
+            // compares
+            '=' => Some((2, ExprType::Int, "Equal")),
+            '>' => Some((2, ExprType::Int, "Greater")),
+            '<' => Some((2, ExprType::Int, "Less")),
+            _ => None,
+        },
+
+        Some(Ident(ref x)) => match x.to_string().as_str() {
+            "and" => Some((1, ExprType::Bool, "And")),
+            "or"  => Some((1, ExprType::Bool, "Or")),
+            _ => None,
+        },
+
+        _ => None,
+    }
+}
+
+/// precedence walking
+fn parse_binop(it: &mut Peekable<IntoIter>, min_prec: i32) -> ParsedExpr {
+    let mut lhs = parse_literal(it);
+    while let Some(prec) = precedence(it) {
+        it.next();
+        if prec.0 < min_prec { break }
+
+        let op = proc_macro2::TokenStream::from_str(prec.2).unwrap();
+        let rhs = parse_literal(it);
+
+        if lhs.t != prec.1 || rhs.t != prec.1 {
+            panic!("error: cannot perform binary operator on these!");
+        }
+
+        let lhs_tokens = lhs.tokens;
+        let rhs_tokens = rhs.tokens;
+        lhs = ParsedExpr{ t: prec.1, tokens: quote!{ #op::new(#lhs_tokens, #rhs_tokens) }};
+    }
+
+    lhs
+}
+
+fn is_ident(a: &TokenTree, b: &'static str) -> bool {
```

Comment:
Returning to the `syn` comment already made: this functionality is provided by `syn::Path::is_ident`.